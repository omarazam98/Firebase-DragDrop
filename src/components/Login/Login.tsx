import React, { Component } from 'react';
import { withAPI } from '@winwin/api-firebase';
import { withRouter } from 'react-router-dom';


interface LoginState {
    email: string;
    password: string;
    persistAuth: boolean;
}

const INITIAL_STATE: LoginState = {
    email: '',
    password: '',
    persistAuth: false,
};

export class Login extends Component<any, LoginState> {
    constructor(props) {
        super(props);
        this.state = INITIAL_STATE;
        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleChange(e){
        e.persist()
        if (Object.keys(this.state).includes(e.target.name)) {
            const value = e.target.type === 'checkbox' ? e.target.checked : e.target.value;
            this.setState(function (prevState, props) {
                return {
                    [e.target.name]: value
                } as Pick<LoginState, keyof LoginState>
            });
        }
        ;
    };

    handleSubmit(e) {
        e.preventDefault();
        return this.props.api.api.auth.firebaseAuth.setPersistence(this.state.persistAuth ? 'local' : 'session').then(() => {
            return this.props.api.api.auth.login.signIn(this.state.email, this.state.password)
                .then((userCredential) => {
                    if (userCredential.user.emailVerified) {
                        this.props.history.push('/dashboard');
                    } else {
                        this.props.history.push('/emailverificationrequired');
                    }
                }).catch((error) => {
                    alert(error.message)
                });
        });
    }


    render() {
        return (
            <div>
                <h1> Login Page </h1>
                <form onSubmit={this.handleSubmit}>

                    <label htmlFor="email">Email: </label> <br/>
                    <input name="email" id="email"
                           value={this.state.email}
                           onChange={this.handleChange}
                           type="email"
                           placeholder="Email Address"/><br/>

                    <label htmlFor="password">Password: </label><br/>
                    <input name="password"
                           id="password"
                           value={this.state.password}
                           onChange={this.handleChange}
                           type="password"
                           placeholder="Password"/><br/>
                    Stay Logged In?
                    <input
                        name="persistAuth"
                        type="checkbox"
                        onChange={this.handleChange} /><br/>

                    <input id={'submitButton'}
                           type="submit"
                           value={'Log In'}></input>
                </form>
            </div>
        );
    };
}

export default withAPI(withRouter(Login))
