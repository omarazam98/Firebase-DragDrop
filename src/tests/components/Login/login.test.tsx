import React from 'react';
import ReactDOM from 'react-dom';
import { Login } from '../../../components/Login/Login';
import { shallow } from 'enzyme';

let wrapper;
let fakeEvent;
let historyMock;
let fakeAPI;
let testUsers;

beforeAll(function (){
    testUsers = {
        verified: {
            email: 'verifiedemail@test.com',
            password: 'verifiedPW'
        },
        unverified: {
            email: 'unverifiedemail@test.com',
            password: 'unverifiedPW'
        },
        invalid: {
            email: 'invalidemail@test.com',
            password: 'invalidPW'
        }
    }

    fakeEvent = { preventDefault: () => {} };
    historyMock = { push: jest.fn() };
    fakeAPI = {
        auth: {
            login: {
                setPersistence: jest.fn().mockResolvedValue(undefined),
                signIn: (email, password) => {
                    return new Promise((resolve, reject) => {
                        if (email === '' || password === '') {
                            reject('Empty fields');
                        }
                        if (email !== testUsers.verified.email && email !== testUsers.unverified.email){
                            reject('No registered user with that email')
                        }else {
                            if(email === testUsers.verified.email && password !== testUsers.verified.password){
                                reject("Wrong password")
                            }
                            const userCredential = {
                                user: {
                                    name: 'sample user',
                                    emailVerified: email === testUsers.verified.email,
                                }
                            }
                            resolve(userCredential);
                        }
                    });
                },
            },
            signOut: jest.fn(),
        }
    };
})

beforeEach(() => {
    wrapper = shallow(<Login />);
});

test('has a valid snapshot', () => {
    expect(wrapper).toMatchSnapshot();
});

test('redirects to dashboard after submit if email verified', () => {
    const historyWrapper = shallow(<Login history={historyMock} api={fakeAPI}/>)
    historyWrapper.setState({email: testUsers.verified.email, password: testUsers.verified.password});
    historyWrapper.instance().handleSubmit(fakeEvent).then(()=>{
        expect(historyMock.push.mock.calls[0]).toEqual([ '/dashboard' ]);
    })
});

test('redirects to email verify after submit if email not verified', () => {
    const historyMock = { push: jest.fn() };
    const historyWrapper = shallow(<Login history={historyMock} api={fakeAPI}/>)
    historyWrapper.setState({email: testUsers.unverified.email, password: testUsers.unverified.password});
    historyWrapper.instance().handleSubmit(fakeEvent).then(()=>{
        expect(historyMock.push.mock.calls[0]).toEqual([ '/email' ]);
    })
});

test('failed log in with empty fields', (done) => {
    const apiWrapper = shallow(<Login api={fakeAPI}/>);
    jest.spyOn(apiWrapper.instance(), 'handleSubmit');
    apiWrapper.instance().handleSubmit(fakeEvent).then(() => {
        expect(apiWrapper.state('error')).toBe('Empty fields');
        done();
    });
});

test('failed log in with non registered email', (done) => {
    const apiWrapper = shallow(<Login api={fakeAPI}/>);
    jest.spyOn(apiWrapper.instance(), 'handleSubmit');
    apiWrapper.setState({email: testUsers.invalid.email, password: testUsers.invalid.password});
    apiWrapper.instance().handleSubmit(fakeEvent).then(() => {
        expect(apiWrapper.state('error')).toBe('No registered user with that email');
        done();
    });
});

test('failed log in with bad password', (done) => {
    const apiWrapper = shallow(<Login api={fakeAPI}/>);
    jest.spyOn(apiWrapper.instance(), 'handleSubmit');
    apiWrapper.setState({email: testUsers.verified.email, password: 'incorrectpassword'});
    apiWrapper.instance().handleSubmit(fakeEvent).then(() => {
        expect(apiWrapper.state('error')).toBe('Wrong password');
        done();
    });
});
