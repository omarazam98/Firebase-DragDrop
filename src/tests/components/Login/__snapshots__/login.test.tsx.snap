// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has a valid snapshot 1`] = `
<div>
  <h1>
     Login Page 
  </h1>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="email"
    >
      Email: 
    </label>
     
    <br />
    <input
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="Email Address"
      type="email"
      value=""
    />
    <br />
    <label
      htmlFor="password"
    >
      Password: 
    </label>
    <br />
    <input
      id="password"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <br />
    Stay Logged In?
    <input
      name="persistAuth"
      onChange={[Function]}
      type="checkbox"
    />
    <br />
    <input
      id="submitButton"
      type="submit"
      value="Log In"
    />
  </form>
</div>
`;
